HEN_HOUSE/egs++/geometry/egs_autoenvelope/volcor.h

```diff
@@ -0,0 +1,663 @@
+        egsInformation(" Inscribed geometry volume   = %.5E cm^3\n", inscribed_volume);
+        options->rng->describeRNG();
+        egsInformation(" -----------------------------------------------\n");
+
+    }
+
+};
+
+
+/*! \brief Apply volume corrections to base regions.
+ *
+ * After the MC volume run is complete, use the tallied hits and mode requested
+ * to set the correct volume of the base regions. */
+vector<EGS_Float> applyVolumeCorrections(VCOptions *opts, HitCounterT hit_counter, vector<EGS_Float> uncorrected) {
+
+    bool zero = opts->mode == ZERO_VOL;
+    double bounds_vol = opts->bounds_volume;
+    double npoints =  opts->npoints;
+    vector<EGS_Float> corrected_vols(uncorrected);
+
+    for (HitCounterT::iterator hi = hit_counter.begin(); hi != hit_counter.end(); hi++) {
+
+        int base_reg = hi->first;
+        if (base_reg < 0) {
+            continue;
+        }
+
+        int hits = hi->second;
+        EGS_Float corrected = uncorrected[base_reg] - bounds_vol*double(hits)/npoints;
+        corrected_vols[base_reg] = zero ? 0 : corrected;
+    }
+
+    return corrected_vols;
+}
+
+/*! \brief Apply file volume corrections to base regions.
+ *
+ * After volumes are loaded from file use the volumes and mode requested
+ * to set the correct volume of the base regions. */
+vector<EGS_Float> applyFileVolumeCorrections(VCOptions *opts, vector<RegVolume> &reg_volumes, vector<EGS_Float> uncorrected) {
+
+    bool zero = opts->mode == ZERO_VOL;
+    vector<EGS_Float> corrected_vols(uncorrected);
+
+    for (size_t rvc=0; rvc < reg_volumes.size(); rvc++) {
+        int base_reg = reg_volumes[rvc].first;
+        EGS_Float vol =  reg_volumes[rvc].second;
+        corrected_vols[base_reg] = zero ? 0 : vol;
+    }
+
+    return corrected_vols;
+}
+
+vector<EGS_Float> getUncorrectedVolumes(EGS_BaseGeometry *base) {
+    vector<EGS_Float> uncorrected;
+    for (int ir=0; ir < base->regions(); ir++) {
+        EGS_Float vol = base->getMass(ir)/base->getRelativeRho(ir);
+        uncorrected.push_back(vol);
+    }
+    return uncorrected;
+
+}
+
+/*! \brief Run the MC simulation.
+ *
+ * Finds regions with inscribed geometries and corrects
+ * the volume of those regions. The algorithm is described
+ * on main page of the Auto Envelope documentation. */
+VCResults findRegionsWithInscribed(VCOptions *opts, EGS_BaseGeometry *base,
+                                   vector<EGS_BaseGeometry *> inscribed, vector<EGS_AffineTransform *> transforms) {
+
+    EGS_Timer timer;
+    timer.start();
+
+    VCResults results(opts);
+
+    results.uncorrected_volumes =  getUncorrectedVolumes(base);
+
+    EGS_Vector point;
+
+    EGS_I64 n_in_inscribed = 0;
+
+    //use first inscribed object to test if point is in inscribed geom
+    EGS_BaseGeometry *first_inscribed = inscribed[0];
+    EGS_AffineTransform *first_transform = transforms[0];
+
+    HitCounterT hit_counter;
+
```

Comment:
Sounds good to me! We can revisit this section of the code when we look at integrating the other volume calculation algorithm (Fred is working on improving it further first, and there is a bug in the current PR).